src/simulation/simulation.jl

```diff
@@ -499,76 +578,108 @@ function _apply_warm_start!(model::DecisionModel)
     return
 end
 
+function _update_system_state!(sim::Simulation, model_name::Symbol)
+    sim_state = get_simulation_state(sim)
+    system_state = get_system_states(sim_state)
+    decision_state = get_decision_states(sim_state)
+    simulation_time = get_current_time(sim_state)
+    for key in get_dataset_keys(decision_state)
+        state_data = get_dataset(decision_state, key)
+        last_update = get_update_timestamp(decision_state, key)
+        simulation_time > get_end_of_step_timestamp(state_data) && continue
+        if last_update <= simulation_time
+            update_system_state!(system_state, key, decision_state, simulation_time)
+        else
+            error("Something went really wrong. Please report this error. \\
+                  last_update: $(last_update) \\
+                  simulation_time: $(simulation_time) \\
+                  key: $(encode_key_as_string(key))")
+        end
+    end
+
+    IS.@record :execution StateUpdateEvent(simulation_time, model_name, "SystemState")
+
+    return
+end
+
+function _update_system_state!(sim::Simulation, model::DecisionModel)
+    _update_system_state!(sim, get_name(model))
+    return
+end
+
+function _update_system_state!(sim::Simulation, model::EmulationModel)
+    sim_state = get_simulation_state(sim)
+    simulation_time = get_current_time(sim)
+    system_state = get_system_states(sim_state)
+    store = get_simulation_store(sim)
+    em_model_name = get_name(model)
+    for key in get_container_keys(get_optimization_container(model))
+        !should_write_resulting_value(key) && continue
+        update_system_state!(system_state, key, store, em_model_name, simulation_time)
+    end
+    IS.@record :execution StateUpdateEvent(simulation_time, get_name(model), "SystemState")
+    return
+end
+
+function _update_simulation_state!(sim::Simulation, model::EmulationModel)
+    _update_system_state!(sim, get_name(model))
+    _update_system_state!(sim, model)
+    return
+end
+
 function _update_simulation_state!(sim::Simulation, model::DecisionModel)
     model_name = get_name(model)
     store = get_simulation_store(sim)
     simulation_time = get_current_time(sim)
     state = get_simulation_state(sim)
-    for field in fieldnames(ValueStates)
-        model_params = get_model_params(store, model_name)
+    model_params = get_decision_model_params(store, model_name)
+    for field in fieldnames(DatasetContainer)
         for key in list_fields(store, model_name, field)
             # TODO: Read Array here to avoid allocating the DataFrame
```

Comment:
Does this TODO still apply? I guess it's tied to my other comment about making a non-copy read_results function.

Reply:
Yes, it does apply. But after testing I wonder how much it matters.